### ISSUE TITLE
Duplicate middlewares in the stack

### ISSUE THREAD

**Issue by esender** [hanami/hanami#638]
Hello!

I have setup Hanami with [omniauth-google-oauth2](https://github.com/zquestz/omniauth-google-oauth2) and experienced weird behavior of callback_request, it was called twice according to console.

After some investigation, I found, that Hanami pushes each middleware [to the `@stack` variable](https://github.com/hanami/hanami/blob/master/lib/hanami/middleware.rb#L66) twice. First time when the server starts, and for second time at request handling. Thus, we have duplicate middlewares in the stack.

Here is naive fix for this issue:

``` ruby
def use(middleware, *args, &blk)
  @stack.push [middleware, args, blk] unless @stack.find { |m| m.first == middleware  }
end
```

This fix has helped me with omniauth.

Is this behavior used for some functionality in Hanami or is this bug?

**Comment by jodosha** [comment]
@esender Hi, can you please paste here (or gist) the content of `@stack`? Thanks!

---

I'm afraid that the uniqueness must be consider all the values (`middleware`, `args` and `blk`) and not just the `middleware`.

Let me give you a naive example: imagine a Rack middleware for logging. We want to mount multiple instances, but with different log levels passed as argument.

```ruby
use MyLoggingMiddleware, :debug
use MyLoggingMiddleware, :notice
```

With that proposed fix, only the first instance will be added to the stack. What we want here is to add both of them.

Probably, `@stack` should not be an `Array`, but a `Set`, which guarantees uniqueness for free.

**Comment by esender** [comment]
My fix is only a naive solution and is not for production use.

I have prepared a simple example to show how to reproduce the bug in this [repository](https://github.com/esender/duplicate_middlewares_example) with description of required steps in [README](https://github.com/esender/duplicate_middlewares_example/blob/master/README.md). This repo contains only Hanami and simple [middleware](https://github.com/esender/duplicate_middlewares_example/blob/master/lib/middlewares/example.rb).

I have add `@stack` printing to the `use` method in Hanami::Middleware. That's what console looks like after the server start:

```
[[Example, [], nil]]
```

This is output after sending request to the server:

```
[[Example, [], nil], [Example, [], nil]]
[[Example, [], nil], [Example, [], nil], [Hanami::Welcome, [], nil]]
```

Would it probably be correct decision to use push middleware only once, when server starts or when request is received? This would prevent extra middleware push to `@stack`.

**Comment by jodosha** [comment]
@esender Here's my output, isn't this the expected behavior?

```shell
[2016-08-16 10:15:04] INFO  WEBrick 1.3.1
[2016-08-16 10:15:04] INFO  ruby 2.3.1 (2016-04-26) [x86_64-darwin15]
[2016-08-16 10:15:04] INFO  WEBrick::HTTPServer#start: pid=37841 port=2300
Run middleware 1
::1 - - [16/Aug/2016:10:15:16 +0200] "GET / HTTP/1.1" 200 - 0.0375
Run middleware 1
::1 - - [16/Aug/2016:10:15:24 +0200] "GET / HTTP/1.1" 200 - 0.0309
Run middleware 1
::1 - - [16/Aug/2016:10:15:44 +0200] "GET / HTTP/1.1" 200 - 0.0328
```

**Comment by esender** [comment]
I still have this issue in my environment. I've tested with OSX and Ubuntu, both ruby 2.3.0 and 2.3.1
I’ve also asked my friend to setup this test project, he got the same result.
Could I ask you to double check it again?

**Comment by fidothe** [comment]
Hey, we just came across this (it's new in 0.8). It's actually a symptom of a deeper underlying issue: The configuration block is being executed twice, so `middleware.use` gets called twice. I've opened a separate issue over at #643 for this

**Comment by jodosha** [comment]
Fixed by https://github.com/hanami/hanami/commit/115b3302a20bc7717538d1d1580ea62a66ed3a1d

**Comment by beauby** [comment]
Should we close this then?

**Comment by joneslee85** [comment]
@esender can you please verify if the fix works for you?

**Comment by jodosha** [comment]
As #657 was merged, we can close this too.
